server: first pass at splitting up NewServer

tbg opened the pull request:
First four commits from #46843.

----

Ultimately we want to separate the SQL and KV startup paths.
This is difficult because of various cyclic dependencies;
for example, the KV subsystem relies on being able to run
simple SQL queries, and the SQL subsystem in turn relies
on node liveness, Gossip, and others that in a multi-tenant
future it won't have access to.

This won't be untangled in one commit, but here is a modest start: We
split out a `newSQLServer` method which we call in `NewServer` and which
populates the `(Server).sqlServer` field, which now houses all that was
formerly in `Server` but really ought to be owned by SQL.

A few of these PRs in, you can imagine a structure emerging in which,
modulo the cross-dependencies which will then clearly be visible,

```
type Server struct
  // ...
  kvServer  *kv.Server
  sqlServer *sql.Server
}
```

Release note: None

tbg committed:
server: first pass at splitting up NewServer

Ultimately we want to separate the SQL and KV startup paths.
This is difficult because of various cyclic dependencies;
for example, the KV subsystem relies on being able to run
simple SQL queries, and the SQL subsystem in turn relies
on node liveness, Gossip, and others that in a multi-tenant
future it won't have access to.

This won't be untangled in one commit, but here is a modest start: We
split out a `newSQLServer` method which we call in `NewServer` and which
populates the `(Server).sqlServer` field, which now houses all that was
formerly in `Server` but really ought to be owned by SQL.

A few of these PRs in, you can imagine a structure emerging in which,
modulo the cross-dependencies which will then clearly be visible,

```
type Server struct
  // ...
  kvServer  *kv.Server
  sqlServer *sql.Server
}
```

Release note: None

tbg committed:
server: clarify dependencies

Explain all of the objects which are passed to newSQLServer for which
this isn't abundantly clear.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46975)
<!-- Reviewable:end -->

tbg committed:
server: avoid fragile pattern in NewServer

This pattern is bad:

```go
func NewServer() *Server {
  s.a = newA()
  s.b = newB(s.a, s.c)
  s.c = newC()
  return s
}
```

Note how `newB` gets called with `s.c == nil`. Creating a partially
populated `Server` struc early is not a good idea; there is no
protection against using fields before they are populated.

Instead, use this pattern:

```go
func NewServer() *Server {
  a := newA()
  c := newC()
  b := newB(a, c)
  return &Server{a: a, b: b, c: c}
}
```

Note how the bug from the first example would be caught at compile time
because we can't use something before it's defined.

In fact, carrying out this refactor revealed a bug: we were always
passing a `nil` engines slice to `NewNodeLiveness`, in effect disabling
the disk stall protection in #32978.

Release note: None

tbg committed:
server: don't shadow package name

This compiles and works fine, but it's not good.

Release note: None

tbg committed:
server: clarify when which fields are initialized

Release note: None

tbg committed:
server: avoid fragile pattern in newSQLServer

Give newSQLServer the same treatment given to NewServer a few commits
ago.

Release note: None

tbg committed:
server: simplify initServer usage

Previously, initServer was initiated in NewServer and did all of its
real setup in ServeAndWait(). This required cumbersome manual
synchronization between incoming Bootstrap requests and the call to
ServeAndWait.

It turns out that we can register the init server with grpc later. All
that matters is that it precedes the first call to `grpc.Serve`. The
code was rearranged to make this true, at which point we could decide
whether to even allow bootstrap before registering the initServer with
grpc. As a result, everything is now more straightforward.

There still seems to be something we should do to clarify when which
listener opens and what gets served, but that is left to further
commits.

Release note: None

tbg committed:
server: improve a comment

Release note: None

andreimatei reviewed (approved):
LGTM

:fist: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-:-M4LIzsg0YPSYDWHWwNy:btn9nxn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @tbg)

___
*[pkg/kv/kvserver/client_test.go, line 872 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L5-d6BFkqhxQiUUAH:-M4L5-d6BFkqhxQiUUAI:b-jlg64f) ([raw file](https://github.com/cockroachdb/cockroach/blob/497a15255a02a1fe06e6c62b6a903e115126e16c/pkg/kv/kvserver/client_test.go#L872)):*
> ```Go
> 	m.populateDB(idx, cfg.Settings, stopper)
> 	nlActive, nlRenewal := cfg.NodeLivenessDurations()
> 	m.nodeLivenesses[idx] = kvserver.NewNodeLiveness(
> ```

I think this should be in the previous commit?
___
*[pkg/kv/kvserver/node_liveness.go, line 172 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L3y133QmDvEfqri1G:-M4L3y140w40vG3j9UFq:b-j380h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f62b1f82ab66800220a3bb54b585f94b7ebde60/pkg/kv/kvserver/node_liveness.go#L172)):*
> ```Go
> 		nodes             map[roachpb.NodeID]storagepb.Liveness
> 		heartbeatCallback HeartbeatCallback
> 		engines           []storage.Engine
> ```

mind taking the opportunity to comment?
___
*[pkg/kv/kvserver/node_liveness.go, line 420 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L4GEXCg5Z2WU-7Qhe:-M4L4GEXCg5Z2WU-7Qhf:be8csup) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f62b1f82ab66800220a3bb54b585f94b7ebde60/pkg/kv/kvserver/node_liveness.go#L420)):*
> ```Go
> // HeartbeatCallback will be invoked whenever this node updates its own liveness.
> func (nl *NodeLiveness) StartHeartbeat(
> 	ctx context.Context, stopper *stop.Stopper, alive HeartbeatCallback, engines []storage.Engine,
> ```

pls comment the `engines` arg here too.
___
*[pkg/server/init.go, line 226 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L90wR8Nv7iZMyU0hL:-M4L90wR8Nv7iZMyU0hM:bxoov8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/init.go#L226)):*
> ```Go
> 	s.mu.Unlock()
> 
> 	// NB: this isn't necessary since bootstrapCluster would fail, but this is
> ```

I'd take the opportunity to get rid of this note. It seems like a silly note to me.
___
*[pkg/server/server.go, line 347 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KmgNTBX_OACf1usr4:-M4KmgNTBX_OACf1usr5:bmcc2ec) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a0f093aaef512609af1f1408c6f834a8d58dc16/pkg/server/server.go#L347)):*
> ```Go
> 	serverpb.RegisterInitServer(s.grpc.Server, s.initServer)
> 
> 	s.runtime = status.NewRuntimeStatSampler(context.TODO(), s.clock)
> ```

everybody else around here uses `context.Background()`...
___
*[pkg/server/server.go, line 614 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KoDrDBTNQoV7_kH_V:-M4KoDrDBTNQoV7_kH_W:bkjbdt8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a0f093aaef512609af1f1408c6f834a8d58dc16/pkg/server/server.go#L614)):*
> ```Go
> 
> type sqlServerArgs struct {
> 	*Config
> ```

hmm this `Config` here seems weird to me. Could we split that guy into a sql-part and a non-sql part?
___
*[pkg/server/server.go, line 807 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4Kr0umCLHJGkmuYMtc:-M4Kr0umCLHJGkmuYMtd:bxbm435) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a0f093aaef512609af1f1408c6f834a8d58dc16/pkg/server/server.go#L807)):*
> ```Go
> 	}
> 
> 	ctx := context.TODO()
> ```

TODO in a new method? Take a ctx. Below too.
___
*[pkg/server/server.go, line 617 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KpbydBAsXgu01qiyq:-M4KpbydBAsXgu01qiyr:b-4in1qy) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L617)):*
> ```Go
> 	*Config
> 	stopper    *stop.Stopper
> 	clock      *hlc.Clock
> ```

can you give a hint as to why SQL needs the HLC clock?
___
*[pkg/server/server.go, line 618 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KplLq9neTWQyhVqOn:-M4KplLq9neTWQyhVqOo:b-3tszio) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L618)):*
> ```Go
> 	stopper    *stop.Stopper
> 	clock      *hlc.Clock
> 	rpcContext *rpc.Context
> ```

say that this is for DistSQL.
___
*[pkg/server/server.go, line 630 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4Kt8zKB7_yKtnz-zoY:-M4Kt8zL4jCDrK8Y8lVN:b676cg) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L630)):*
> ```Go
> 	nodeLiveness *kvserver.NodeLiveness
> 	// The executorConfig uses the provider.
> 	protectedtsProvider protectedts.Provider
> ```

since you're putting this commenting effort in, wanna pierce through the `ExecutorConfig` vail and see what the different pieces are actually used for?
___
*[pkg/server/server.go, line 649 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4Kq9PS5QOkraV9uquN:-M4Kq9PS5QOkraV9uquO:b-oeeoyx) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L649)):*
> ```Go
> 	// SQL needs to use KV. There's one regular DB and one for DistSQL.
> 	//
> 	// TODO(tbg): why can't it be the same?
> ```

It should indeed be the same. The differences in how we initialize them don't seem important. Let's unify them!
___
*[pkg/server/server.go, line 658 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KripN7JOLtcvw7q1z:-M4KripN7JOLtcvw7q2-:bqp7zkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L658)):*
> ```Go
> 	// KV depends on the internal executor, so it is early bound but only filled
> 	// in newSQLServer.
> 	internalExecutor *sql.InternalExecutor // empty initially
> ```

Remove the "empty initially". 

How about replacing this pointer with a callback (or a `container` struct that multiple callbacks can be registered on)? This pointer business is confusing.
___
*[pkg/server/server.go, line 661 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KssMpCriKs36NMIdy:-M4KssMq-EWNC2D28Z1F:b34et41) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L661)):*
> ```Go
> 	// DistSQL, lease management, and others want to know the node they're on.
> 	//
> 	// TODO(tbg): reasonable ask, but a SQL tenant process has no NodeID.
> ```

nit: I wouldn't put this TODO in... Doesn't explain anything, and so it doesn't help anybody.
___
*[pkg/server/server.go, line 403 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KtwRU2f0H39Ac4LxZ:-M4KtwRU2f0H39Ac4Lx_:b8tpp6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/613316c806943a27faed4042d468c1c48f3bf414/pkg/server/server.go#L403)):*
> ```Go
> 		// TODO(tbg): this is supposed to get the engines slice, which is only
> 		// available at start time.
> 		nil,
> ```

nit: put the argument name

And I don't quite understand the TODO. Are you saying that the ctor signature is bad?

Nvm, I see that it went away indeed later.
___
*[pkg/server/server.go, line 555 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L3Z2W87xIbgrD6Zcu:-M4L3Z2W87xIbgrD6Zcv:b-69ju9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/613316c806943a27faed4042d468c1c48f3bf414/pkg/server/server.go#L555)):*
> ```Go
> 
> 	lateBoundServer := &Server{}
> 	// TODO(tbg): don't pass the whole Server into lateBoundServer to avoid this
> ```

You mean making `newAdminServer()` take more granular dependencies? I think this TODO a bit unclear.
___
*[pkg/server/server.go, line 696 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L5N_t7BvELA5G6OEZ:-M4L5N_t7BvELA5G6OE_:b7924o3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddf877af1ad4f91fe9c482eb4ab5b747f2939f48/pkg/server/server.go#L696)):*
> ```Go
> 	// KV depends on the internal executor, so it is early bound but only filled
> 	// in newSQLServer.
> 	lateBoundInternalExecutor *sql.InternalExecutor // empty initially
> ```

I think we need to decide what "late" and "early" bound mean. Now the comment says "early" but the name says "late" :P

I stand by my previous suggestion to encapsulate the dissemination of an initialized `executor` into a more explicit struct.
___
*[pkg/server/server.go, line 1506 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LFmPT8aUsdQSqEL0u:-M4LFmPT8aUsdQSqEL0v:b-3eyesh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1506)):*
> ```Go
> 	}
> 
> 	// We cannot register a server after grpc.Serve has been called (it will
> ```

nit: can you clarify this comment but spelling out that we would like to register the initServer later?
But actually, would we even want to register the service later? The structure of the code that you've introduced seems fine to me - the `initServer` is nicely turned into a dummy if the cluster had been bootstrapped.
___
*[pkg/server/server.go, line 1673 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LHIYK0Y-4yZ_ok9_q:-M4LHIYL8ltkvYnKmbe4:bnlnr6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1673)):*
> ```Go
> 
> 	// This opens the main listener.
> 	startRPCServer(workersCtx)
> ```

can we move this down below the `if len(s.cfg.GossipBootstrapResolvers) == 0 {` stanza? It'd be closer to the `ServeAndWait()` call, and they seem related.
___
*[pkg/server/server.go, line 1679 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LHBOaC22KWnBMIDBu:-M4LHBOb5xhKc1vXBidO:b-hcox96) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1679)):*
> ```Go
> 		// Note that we're not checking whether the node is already bootstrapped;
> 		// if this is the case, Bootstrap will simply fail.
> 		_ = s.stopper.RunAsyncTask(ctx, "bootstrap", func(ctx context.Context) {
> ```

how come this is done async?
___
*[pkg/server/server.go, line 1691 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LIHwN78BqWo909AyM:-M4LIHwN78BqWo909AyN:bk0261m) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1691)):*
> ```Go
> 	}
> 
> 	state, err := initServer.ServeAndWait(ctx, s.stopper, s.gossip)
> ```

It kinda bothers me that we have the `Bootstrap()` call above, but also this `ServeAndWait()` call unconditionally.
What do you think of either:
- make `Bootstrap()` wait for gossip to be connected, and return the `initState`. And then don't call `ServeAndWait()` if we bootstrapped. Or,
- move the `if len(s.cfg.GossipBootstrapResolvers) == 0` check into `initServer.ServeAndWait()`. This would also eliminate, or at least hide, the ugly handling of `ErrClusterInitialized`.


<!-- Sent from Reviewable.io -->

tbg committed:
server: self-bootstrap without async task

The async task was not necessary. Bootstrap() writes to a one-buffered
channel that is only written to once.

Release note: None

tbg committed:
server: self-bootstrap before opening listener

This looks nicer and also means that self-bootstrap will always get
there first (which doesn't matter but is nice, too).

Release note: None

tbg committed:
server: don't pass nil engines to NewNodeLiveness

Set things up so that we *actually* check for disk stalls before
liveness updates.

Release note (bug fix): A mechanism to detect stalled disks was
erroneously left disabled. It is now enabled: nodes will verify that
they can write to their engines before each attempt to announce
themselves as live to the cluster.

tbg committed:
kvserver: improve liveness commentary

Release note: None

tbg committed:
server: improve Bootstrap() error behavior

Just make it a little more obvious that either

a) the first bootstrap is already refused with ErrClusterInitialized
b) the first bootstrap catches an error, errInternalBootstrapError
c) it goes through

Any additional attempt will receive either ErrClusterInitialized(after
case a) or c)) or errInternalBootstrapError (case b).

Release note: None

tbg committed:
server: avoid context.TODO()

I had sprinkled some of them and forgotten to replace them, and some of
them predate my work. Either way, all are gone now.

Release note: None

tbg committed:
server: improve some comments

Release note: None

tbg committed:
server: use same `kv.DB` for SQL and DistSQL

They're set up the same, except that one had a better context
annotation.

Release note: None

tbg committed:
server: clarify internal executor initialization

Release note: None

tbg committed:
server: improve TODO about NodeID in multi-tenancy

Release note: None

tbg committed:
server: move/reword comment about registering initServer

Release note: None

tbg reviewed (commented):
TFTR! I agree that there's more to be done and that is the plan.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-:-M4PB4I_BT-4NnScosKR:b80eyqz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvserver/client_test.go, line 872 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L5-d6BFkqhxQiUUAH:-M4P2o8Y3Q9TH7PENSNe:bw10jn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/497a15255a02a1fe06e6c62b6a903e115126e16c/pkg/kv/kvserver/client_test.go#L872)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this should be in the previous commit?
</blockquote></details>

Oops, done.
___
*[pkg/kv/kvserver/node_liveness.go, line 172 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L3y133QmDvEfqri1G:-M4P37LpBH9cCPAyKpdS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f62b1f82ab66800220a3bb54b585f94b7ebde60/pkg/kv/kvserver/node_liveness.go#L172)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

mind taking the opportunity to comment?
</blockquote></details>

Done.
___
*[pkg/kv/kvserver/node_liveness.go, line 420 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L4GEXCg5Z2WU-7Qhe:-M4P3GOX1PjfB42B_jL8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f62b1f82ab66800220a3bb54b585f94b7ebde60/pkg/kv/kvserver/node_liveness.go#L420)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls comment the `engines` arg here too.
</blockquote></details>

Done.
___
*[pkg/server/init.go, line 226 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L90wR8Nv7iZMyU0hL:-M4P4sUoBqVv5XJ7OpP9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/init.go#L226)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'd take the opportunity to get rid of this note. It seems like a silly note to me.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 347 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KmgNTBX_OACf1usr4:-M4P5rOb6QGdGPZStPUY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a0f093aaef512609af1f1408c6f834a8d58dc16/pkg/server/server.go#L347)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

everybody else around here uses `context.Background()`...
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 614 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KoDrDBTNQoV7_kH_V:-M4P6MVuD9g3zaP4XrB0:b17gi14) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a0f093aaef512609af1f1408c6f834a8d58dc16/pkg/server/server.go#L614)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

hmm this `Config` here seems weird to me. Could we split that guy into a sql-part and a non-sql part?
</blockquote></details>

Absolutely, but Rome wasn't built in a day. There will be more PRs.
___
*[pkg/server/server.go, line 807 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4Kr0umCLHJGkmuYMtc:-M4P8X3i26G5XcjFjkEn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a0f093aaef512609af1f1408c6f834a8d58dc16/pkg/server/server.go#L807)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

TODO in a new method? Take a ctx. Below too.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 617 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KpbydBAsXgu01qiyq:-M4P6vgg8qoy-Nu5r5Bv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L617)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can you give a hint as to why SQL needs the HLC clock?
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 618 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KplLq9neTWQyhVqOn:-M4P6w3m1652Ci57FGrg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L618)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

say that this is for DistSQL.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 630 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4Kt8zKB7_yKtnz-zoY:-M4P6yrjD3ApUt2v2PTU:b-hlvfvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L630)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

since you're putting this commenting effort in, wanna pierce through the `ExecutorConfig` vail and see what the different pieces are actually used for?
</blockquote></details>

Generally yes, in this PR no.
___
*[pkg/server/server.go, line 649 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4Kq9PS5QOkraV9uquN:-M4P7Sv--gQwk657pUq_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L649)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It should indeed be the same. The differences in how we initialize them don't seem important. Let's unify them!
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 658 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KripN7JOLtcvw7q1z:-M4P7tBf015cWUo5CZvY:btgjv1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L658)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Remove the "empty initially". 

How about replacing this pointer with a callback (or a `container` struct that multiple callbacks can be registered on)? This pointer business is confusing.
</blockquote></details>

Renamed to `circularInternalExecutor` with a specific comment on how it's used. I don't want to build machinery for this yet because this dependency as it stands should not last anyway.
___
*[pkg/server/server.go, line 661 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4KssMpCriKs36NMIdy:-M4P8O86397EsgreLgSw:by6lp7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/47a94d9d6790936a94a462834197bed6e5c314c4/pkg/server/server.go#L661)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: I wouldn't put this TODO in... Doesn't explain anything, and so it doesn't help anybody.
</blockquote></details>

I improved the wording. It's a real TODO (i.e. one that needs to get done). I hope to get SQL to help with some of this work, so these reminders are helpful.
___
*[pkg/server/server.go, line 555 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L3Z2W87xIbgrD6Zcu:-M4P6KwB3lQk-4Iz2JiC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/613316c806943a27faed4042d468c1c48f3bf414/pkg/server/server.go#L555)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

You mean making `newAdminServer()` take more granular dependencies? I think this TODO a bit unclear.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 696 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4L5N_t7BvELA5G6OEZ:-M4P7zmRBYtH5fz2DMMI:b-bang1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddf877af1ad4f91fe9c482eb4ab5b747f2939f48/pkg/server/server.go#L696)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think we need to decide what "late" and "early" bound mean. Now the comment says "early" but the name says "late" :P

I stand by my previous suggestion to encapsulate the dissemination of an initialized `executor` into a more explicit struct.
</blockquote></details>

Answered above.
___
*[pkg/server/server.go, line 1506 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LFmPT8aUsdQSqEL0u:-M4P8gW891VEkFitRlxI:bxq0370) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1506)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: can you clarify this comment but spelling out that we would like to register the initServer later?
But actually, would we even want to register the service later? The structure of the code that you've introduced seems fine to me - the `initServer` is nicely turned into a dummy if the cluster had been bootstrapped.
</blockquote></details>

The code is fine now, but I can already see that there will be temptation to split this method, in which case we'd set up all the listen/serve stuff early. At that point we'll run into trouble, I just don't want there to be confusion when it happens (and ideally remind folks of the constraint early).

Clarified the comment.
___
*[pkg/server/server.go, line 1673 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LHIYK0Y-4yZ_ok9_q:-M4P0xF21RInIGknqYep:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1673)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can we move this down below the `if len(s.cfg.GossipBootstrapResolvers) == 0 {` stanza? It'd be closer to the `ServeAndWait()` call, and they seem related.
</blockquote></details>

Done.
___
*[pkg/server/server.go, line 1679 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LHBOaC22KWnBMIDBu:-M4P0a5lDVTRXYEpyAPg:bemm3au) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1679)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how come this is done async?
</blockquote></details>

The channel that communicates the bootstrap result to ServeAndWait was unbuffered. I made it buffered so we can avoid the async task here.
___
*[pkg/server/server.go, line 1691 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/46975#-M4LIHwN78BqWo909AyM:-M4P0GgO5SRncW0O4RYl:bjo08y1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce7f6b58975b95f14c27e0bb52b03ec0a3f7b25e/pkg/server/server.go#L1691)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

It kinda bothers me that we have the `Bootstrap()` call above, but also this `ServeAndWait()` call unconditionally.
What do you think of either:
- make `Bootstrap()` wait for gossip to be connected, and return the `initState`. And then don't call `ServeAndWait()` if we bootstrapped. Or,
- move the `if len(s.cfg.GossipBootstrapResolvers) == 0` check into `initServer.ServeAndWait()`. This would also eliminate, or at least hide, the ugly handling of `ErrClusterInitialized`.
</blockquote></details>

The fact that ServeAndWait is called unconditionally is a feature as it abstracts away the method by which we arrive at a bootstrapped cluster.

We are deprecating self-bootstrap, so code related to that will go away. This means just deleting this `len(s.cfg.GossipBootstrapResolvers)` block. I am hesitant to do anything with it that mixes it in with the surrounding code more for that reason.


<!-- Sent from Reviewable.io -->

tbg commented:
bors r=andreimatei
